src/main/java/org/folio/circulation/domain/validation/ItemLimitValidator.java

```diff
@@ -0,0 +1,119 @@
+package org.folio.circulation.domain.validation;
+
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+import static org.folio.circulation.support.Result.ofAsync;
+import static org.folio.circulation.support.Result.succeeded;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+import org.apache.commons.lang3.StringUtils;
+
+import org.folio.circulation.domain.Item;
+import org.folio.circulation.domain.Loan;
+import org.folio.circulation.domain.LoanAndRelatedRecords;
+import org.folio.circulation.domain.LoanRepository;
+import org.folio.circulation.domain.policy.LoanPolicyRepository;
+import org.folio.circulation.support.Result;
+import org.folio.circulation.support.ValidationErrorFailure;
+
+public class ItemLimitValidator {
+  private final Function<String, ValidationErrorFailure> itemLimitErrorFunction;
+  private final LoanRepository loanRepository;
+  private final LoanPolicyRepository loanPolicyRepository;
+
+  public ItemLimitValidator(Function<String, ValidationErrorFailure> itemLimitErrorFunction,
+    LoanRepository loanRepository, LoanPolicyRepository loanPolicyRepository) {
+
+    this.itemLimitErrorFunction = itemLimitErrorFunction;
+    this.loanRepository = loanRepository;
+    this.loanPolicyRepository = loanPolicyRepository;
+  }
+
+  public CompletableFuture<Result<LoanAndRelatedRecords>> refuseWhenItemLimitIsReached(
+    LoanAndRelatedRecords records) {
+
+    Loan loan = records.getLoan();
+    Integer itemLimit = loan.getLoanPolicy().getItemLimit();
+
+    if (itemLimit == null) {
+      return completedFuture(succeeded(records));
+    }
+
+    return ofAsync(() -> loan.getLoanPolicy().getRuleConditions())
+      .thenComposeAsync(result -> result.failAfter(ruleConditions -> isLimitReached(ruleConditions, records),
+        ruleConditions -> {
+          String message = getErrorMessage(ruleConditions);
+          return itemLimitErrorFunction.apply(String.format("Patron has reached maximum item limit of %d items %s",
+            itemLimit, message));
+        }))
+      .thenApply(result -> result.map(v -> records));
+  }
+
+  private CompletableFuture<Result<Boolean>> isLimitReached(List<String> ruleConditions,
+                                                            LoanAndRelatedRecords records) {
```

Comment:
Personally, I find that trying to maintain lined up parameter names like this is an unnecessary overhead and wastes horizontal space. I would include it on the next line indented like any other line continuation (2 spaces).

Reply:
done